Launch settings fix

jtorres opened the pull request:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/862782
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/918958
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/933588
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/933582
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/934345
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/934367
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/934322
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/918985
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/755485
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/958219

Specs https://microsoft-my.sharepoint.com/:w:/p/mhutch/EfsOmr-aDnJMkI1Dv1QQIrcBd-kon2e8qB9QicNVkq3E-A?e=mYbAcu

Mote: rebased

User Cases:

given a ASP.NET Core project 

- when a new project is created then it should contain a default launchSettings.json profile called like project name and it should be a Default run config
- when a new Run Config is added then it should update launchSettings.json and add a new profile
- when a Run Config is removed then it should update launchSettings.json and remove that profile
- when a run config is modified the it should update launchSettings.json and the profile updated
- when a profile is added in the launchSettings.json then it should add a new Run Config
- when a profile is removed in the launchSettings.json then it should remove that Run Config
- when a profile is updated in the launchSettings.json then it should update the Run Config
- when a project with more than one profiles in the launchSettings.json is loaded then it should create a new Run Config for each profile which type == "Project"

jtorres committed:
[Core] Fixes an issue when dealing with Default run config + adds OnRemoveRunConfiguration

jtorres committed:
[aspnetcore] Keeps sync'ed all RunConfiguration with launchSettings.json profiles

rodrmoya commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -116,27 +152,109 @@ protected override string OnGetDefaultBuildAction (string fileName)
+					} else {
+						var index = Project.RunConfigurations.IndexOf (runConfig);
+						this.Project.RunConfigurations [index] = runConfig;
+					}
```
hmm, not sure what this does, it seems to just set to the item that was retrieved in line 219 again, by retrieving its index and setting it by that index??

jtorres replied:
right somehow I missed a step before `this.Project.RunConfigurations [index] = runConfig` that copies `profile` values to `runconfig`... will fix

rodrmoya commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/ILaunchProfile.cs:
```diff
@@ -5,7 +5,7 @@
 {
-	interface ILaunchProfile
+	public interface ILaunchProfile
 	{
```
Does this need to be public?

jtorres replied:
same, since LaunchProfileData is public, it should be public

mrward replied:
Asking the same question as Rodrigo here - does this and the LaunchProfileData need to be public. Whilst AspNetCoreConfiguration is public the LaunchProfileData is not public on AspNetCoreConfiguration. It seems like we could make LaunchProfileData and ILaunchProfile non-public.

jtorres replied:
ok, set LaunchProfileData to internal

rodrmoya commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileData.cs:
```diff
@@ -10,9 +10,8 @@
 	[JsonObject (MemberSerialization.OptIn)]
-	internal class LaunchProfileData
+	public class LaunchProfileData
 	{
```
Same here, does it need to be public?

jtorres replied:
yes, `LaunchProfileData` is used by AspnNetCoreConfiguration which is public

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -116,27 +152,109 @@ protected override string OnGetDefaultBuildAction (string fileName)
 		{
 			base.Dispose ();
 			FileService.FileChanged -= FileService_FileChanged;
+			foreach (var conf in aspNetCoreRunConfs) {
```
Maybe also clear/null out the dictionary here.

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreRunConfiguration.cs:
```diff
@@ -216,5 +177,7 @@ internal bool UsingHttps ()
 			return false;
 		}
+
+		public void OnSaveRequested () => SaveRequested?.Invoke (this, EventArgs.Empty);
```
Does this have to be public?

Therzok commented:
There's a lot of mixed whitespace here.

jtorres committed:
[aspnetcore] Few fixes:

- Clear dictionary when Dispose
- Properly copies profile values
- Fixes spaces

mrward commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -24,37 +24,51 @@
-		public LaunchProfileProvider (Project project)
+		public LaunchProfileData DefaultProfile {
+			get {
+				if (!Profiles.ContainsKey (defaultNamespace)) { 
```
TryGet?

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -3060,7 +3060,8 @@ internal void OnRunConfigurationsAdded (IEnumerable<SolutionItemRunConfiguration
-
+			foreach (var s in items)
+				ProjectExtension.OnRemoveRunConfiguration (s.Name);
 		}
```
I think we should follow the same pattern as for NotifyItemsRemoved.

Project.NotifyItemsRemoved -> ProjectExtension.OnItemsRemoved -> Project.OnItemsRemoved

Also I think it should pass the full set of items instead of the name of each one individually.

We could also implement the same pattern when a run config is added whilst we are added the pattern for removing.

mrward commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -116,27 +152,109 @@ protected override string OnGetDefaultBuildAction (string fileName)
 				return;
+			
+			lock (profiles) {
+				updating = true;
```
FileService_FileChanged I think runs on the UI thread so we do not need a lock here. A lock would make sense if we were running this on another background thread.

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -37,15 +38,51 @@ namespace MonoDevelop.AspNetCore
+				return aspNetCoreRunConfs [name];
+
+			var profile = new LaunchProfileData ();
+			if (!launchProfileProvider.Profiles.ContainsKey (name)) {
```
TryGetValue here?

jtorres committed:
[aspnetcore] Few fixes:

- Use TyrGetValue for dicts
- remove lock when FileChanged
- Changes OnRemoveRunConfiguration argument

rodrmoya commented:
I wonder if we could easily have a high level test, that created a project, then a couple run configs and built on both configs and checking some condition?

jtorres committed:
[aspnetcore] When loading launchSettings profiles, if exists then assign it to run Config

jtorres committed:
[AspNetCore] Safety updates CurrentProfile and makes sure EnvVariables are updated too

jtorres committed:
[aspnetcore] Adds additional null checks for Dicts + sorts out "Default" profile initialization

jtorres committed:
[Core] Adds missed method OnRemoveRunConfiguration in DefaultMSBuildExt, otherwise, ProjExtension throws NRE

mrward commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -121,22 +154,111 @@ protected override void OnItemReady ()
+						aspNetCoreRunConfiguration.UpdateProfile (profile.Value);
+						this.Project.RunConfigurations [index] = runConfig;
+					}
+				}
```
This code is almost, but not quite, identical to similar code in OnItemReady - can this code be shared instead of duplicating it here?

mrward reviewed (commented):
Everything else looks good to me.

Just two minor questions about public classes and some possible code duplication.

mrward commented:
Adding or removing a profile from launchSettings.json - the main toolbar does not update with this new run configuration. If you add/remove a run config from project options the main toolbar updates.

Adding a new user specific run config adds an entry into the launchSettings.json - not sure what the expected behaviour here. Not sure I understand the mappings here and how things should be synchronised. The spec mentions ignoring user specific run configurations - but not sure if that is just on load or all the time.

jtorres committed:
[aspnetcore] Updates MainToolBar when modifying profiles directly in launchSettings.json

jtorres commented:
I need to run a smoke test after lasts commits

jtorres committed:
[Core] Reverts defaultBlankRunConfiguration re-use

jtorres committed:
[aspnetcore] Makes LaunchProfileData internal

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -121,22 +154,113 @@ protected override void OnItemReady ()
+			foreach (var conf in aspNetCoreRunConfs) {
+				conf.Value.SaveRequested -= Aspnetconf_Save;
+			}
+			aspNetCoreRunConfs.Clear ();
```
No need to clear if we set it to null.

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -121,22 +154,113 @@ protected override void OnItemReady ()
+			if (updating || launchProfileProvider == null)
+				return;
+			foreach (var item in items) {
+				launchProfileProvider.Profiles.TryRemove (item.Name, out var _);
```
```suggestion
				launchProfileProvider.Profiles.TryRemove (item.Name, out _);
```

Therzok reviewed (commented):
LGTM!

slluis reviewed (changes requested):
We need to document test scenarios and use cases.

jtorres committed:
[aspnetcore] Basically, all the RunConfigs created by LaunchSettings profile is shared config (not user specific)

- Refactoring code in orde to share code + makes it more testeable
- Adds a new unit test

mrward commented:
- Could not test everything since reloading the solution with new run configs throws the null reference exception that Sayed reported.

- Seems to be some problems with shared versus user run configs. From the spec the user specific run configs should be ignored but when you create a user specific run config the launchsettings.json file is updated.

- A shared run configuration does not seem to be created for the Default launchSettings.json profile. From the spec I was assuming that would be the case - maybe that is not the case for Default.

jtorres committed:
[aspnetcore] Prevents NRE

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -37,15 +38,47 @@ namespace MonoDevelop.AspNetCore
+
+			var key = name != "Default" ? name : this.Project.DefaultNamespace;
+
+			if (!launchProfileProvider.Profiles.TryGetValue (key, out var _)) {
```
```suggestion
			if (!launchProfileProvider.Profiles.TryGetValue (key, out profile)) {
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -37,15 +38,47 @@ namespace MonoDevelop.AspNetCore
+				profile = launchProfileProvider.AddNewProfile (key);
+				launchProfileProvider.Profiles [key] = profile;
+			} else {
+				profile = launchProfileProvider.Profiles [key];
```
Since we avoided double lookup, we can remove this block.

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -36,27 +38,42 @@ namespace MonoDevelop.AspNetCore
+		public LaunchProfileData DefaultProfile {
+			get {
+				if (!Profiles.ContainsKey (defaultNamespace)) {
+					var defaultProfile = CreateDefaultProfile ();
```
```csharp
return Profiles [defaultNamespace] = CreateDefaultProfile ();
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -36,27 +38,42 @@ namespace MonoDevelop.AspNetCore
 		public IDictionary<string, JToken> GlobalSettings { get; private set; }
 		internal JObject ProfilesObject { get; private set; }
-		internal string launchSettingsJsonPath => Path.Combine (project.BaseDirectory, "Properties", "launchSettings.json");
+		public ConcurrentDictionary<string, LaunchProfileData> Profiles { get; set; }
```
Why can we publicly set the Profiles dictionary? Should this be readonly?
```suggestion
		public ConcurrentDictionary<string, LaunchProfileData> Profiles { get; }
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -69,25 +86,25 @@ public void LoadLaunchSettings ()
-			}
-			catch {
+				return JObject.Parse (File.ReadAllText (LaunchSettingsJsonPath));
+			} catch {
```
```csharp
try {
        using var fs = File.OpenRead (LaunchSettingsJsonPath);
	using var sr = new StreamReader (fs);
	using var reader = new JsonTextReader (sr);
	return JObject.Load (reader);
} catch (JsonReaderException e) {
	LoggingService.LogInternalError ($"Failed to parse Json {LaunchSettingsJsonPath}", e);
}
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -120,10 +137,22 @@ void ReportError (string message, Exception ex)
 		}
 
-		public LaunchProfileData CreateDefaultProfile ()
+		public LaunchProfileData AddNewProfile (string name)
```
Should we do validation if the profile already exists?

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -120,10 +137,22 @@ void ReportError (string message, Exception ex)
+			if (Profiles == null)
+				Profiles = new ConcurrentDictionary<string, LaunchProfileData> ();
+
+			var newProfile = CreateProfile (name);
```
Replace with:
```csharp
return Profiles [name] = CreateProfile (name); 
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -139,9 +168,62 @@ public LaunchProfileData CreateDefaultProfile ()
+			var profiles = new Dictionary<string, LaunchProfileData> {
+				{ defaultNamespace, CreateDefaultProfile () }
+			};
+			Profiles = new ConcurrentDictionary<string, LaunchProfileData> (profiles);
```
Remove the redundant intermediate dictionary, and set the value directly:
```csharp
			Profiles = new ConcurrentDictionary<string, LaunchProfileData>();
			Profiles.TryAdd (defaultNamespace, CreateDefaultProfile ());
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -139,9 +168,62 @@ public LaunchProfileData CreateDefaultProfile ()
+					};
+					project.RunConfigurations.Add (projectRunConfiguration);
+				} else if (runConfig is AspNetCoreRunConfiguration aspNetCoreRunConfiguration) {
+					var index = project.RunConfigurations.IndexOf (runConfig);
```
Validate the return value is `>= 0` or Debug.Assert it if we expect it to always be found:
```csharp
if (index >= 0) ... // do stuff

// or

Debug.Assert (index >= 0, "Expected value to be found");
```

Therzok replied:
I expect it's the latter. Maybe it would be worth refactoring so we get the index of the first run configuration matching the `Name`, and do the `>= 0` check there. I'm wary of race conditions.

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/LaunchProfileProvider.cs:
```diff
@@ -139,9 +168,62 @@ public LaunchProfileData CreateDefaultProfile ()
+					key = project.DefaultNamespace;
+				}
+
+				if (Profiles.TryGetValue (key, out var _))
```
```suggestion
				if (Profiles.TryGetValue (key, out _))
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreProjectExtension.cs:
```diff
@@ -37,15 +38,47 @@ namespace MonoDevelop.AspNetCore
+
+			var aspnetconf = new AspNetCoreRunConfiguration (name, profile);
+			aspnetconf.SaveRequested += Aspnetconf_Save;
+			aspNetCoreRunConfs.Add (name, aspnetconf);
```
Threading concerns? If the value was added already by another thread, we can race here.

Therzok replied:
If we expect this to be called from the main thread only, add a `Runtime.AssertMainThread()` at the top of the method.

Therzok commented:
First review pass, mostly LGTM!

rodrmoya commented:
@monojenkins backport pr-8.3-launch-settings-fix
